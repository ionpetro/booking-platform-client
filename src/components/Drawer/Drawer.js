export default {
  data() {
    return {
      active: false
    };
  },
  computed: {
    element() {
      return this.$refs.element;
    },
    overlay() {
      return this.$refs.overlay;
    }
  },
  mounted() {
    this.addTransitionListener();
    this.addOnClickListener();
  },
  destroyed() {
    window.removeEventListener('transitionend', this.transitionListener, false);
    window.removeEventListener('click', this.clickListener, false);
  },
  methods: {
    addTransitionListener() {
      this.overlay.addEventListener('transitionend', this.transitionListener, false);
    },
    addOnClickListener() {
      this.overlay.addEventListener('click', this.clickListener, false);
    },
    clickListener() {
      this.close();
    },
    transitionListener(event) {
      this.handleZIndex(event);
    },
    /**
     * HandleZIndex is executed when the drawer transition is finished
     * due to the transitionListener. This way, the overlay will go to
     * the back when drawer is closed and overlay opacity hits 0.
     */
    handleZIndex() {
      const opacity = window.getComputedStyle(this.overlay).getPropertyValue('opacity');
      if (opacity <= 0) {
        this.overlay.style.zIndex = -999;
      }
    },
    overlayOpacity(opacity) {
      this.overlay.style.opacity = opacity;
      if (opacity > 0) {
        this.overlay.style.zIndex = 999;
      }
    },
    open() {
      this.overlayOpacity(1);
      this.lock(document.querySelector('html'));
      this.lock(document.querySelector('body'));
      this.active = true;
    },
    close() {
      this.overlayOpacity(0);
      this.unlock(document.querySelector('html'));
      this.unlock(document.querySelector('body'));
      this.active = false;
    },
    lock(element) {
      const updatedElement = element;
      updatedElement.style.overflow = 'hidden';
      updatedElement.style.touchAction = 'none';
    },
    unlock(element) {
      element.style.removeProperty('overflow');
      element.style.removeProperty('touch-action');
    }
  }
};
