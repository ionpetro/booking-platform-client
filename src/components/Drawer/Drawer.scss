.overlay {
  position: fixed;
  z-index: -999;
  left: 0;
  top: 0;
  background: rgba(11, 10, 12, 0.35);
  opacity: 0;
  width: 100%;
  height: 100%;
  transition: opacity 0.5s ease-in-out;
}

.sidebar {
  right: 0;
  z-index: 9999;
  position: fixed;
  will-change: transform;
  height: 100%;
  top: 0;
  transition: transform 0.5s ease-in-out;
  background: #fff;
  width: 780px;
  overflow-y: auto;
  overflow-x: hidden;
  word-wrap: break-word;

  &__close {
    position: absolute;
    display: flex;
    cursor: pointer;
    align-items: center;
    justify-content: center;
    width: 40px;
    height: 40px;
    left: 16px;
    top: 16px;
    border-radius: 50%;
    background: rgba(0, 0, 0, 0.39);
    box-shadow: 0px 4px 10px rgba(109, 109, 109, 0.25);

    &:hover {
      background: rgba(0, 0, 0, 0.9);
    }
  }

  // for tablet and smaller screens
  // the drawer cover the hole screen
  @media only screen and (max-width: 780px) {
    width: 100vw;
  }
}

.openDrawer {
  .sidebar {
    transform: translate3d(0, 0, 0);
  }
}

.closeDrawer {
  .sidebar {
    transform: translate3d(100%, 0, 0);
  }
}
