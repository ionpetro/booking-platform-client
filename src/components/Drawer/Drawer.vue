<template>
  <section :class="active ? 'openDrawer' : 'closeDrawer'">
    <aside class="sidebar" ref="element">
      <div @click="close" class="sidebar__close">
        <img src="@/assets/icons/Remove.svg" alt="Cancel Icon" />
      </div>
      <slot></slot>
    </aside>
    <div class="overlay" ref="overlay"></div>
  </section>
</template>
<script src="./Drawer.js"></script>
<style lang="scss" src="./Drawer.scss"></style>
