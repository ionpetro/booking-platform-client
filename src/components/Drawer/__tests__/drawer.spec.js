import {shallowMount} from '@vue/test-utils';
import Drawer from '../Drawer.vue';

const wrapper = shallowMount(Drawer, {
  methods: {
    open: jest.fn(),
    addOnKeyPressListener: jest.fn(),
    addOnClickListener: jest.fn()
  }
});

describe('Drawer', () => {
  describe('when loaded', () => {
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
