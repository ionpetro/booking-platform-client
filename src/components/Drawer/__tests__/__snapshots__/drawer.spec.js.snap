// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer when loaded renders correctly 1`] = `
<section
  class="closeDrawer"
>
  <aside
    class="sidebar"
  >
    <div
      class="sidebar__close"
    >
      <img
        alt="Cancel Icon"
        src="@/assets/icons/Remove.svg"
      />
    </div>
     
  </aside>
   
  <div
    class="overlay"
  />
</section>
`;
